[RISCV][MC] Add MC support of Zibi experimental extension

BoyaoWang430 opened the pull request:
This adds the MC support of Zibi v0.1 experimental extension.

References:
* https://lf-riscv.atlassian.net/wiki/spaces/USXX/pages/599261201/Branch+with+Immediate+Zibi+Ratification+Plan
* https://lf-riscv.atlassian.net/browse/RVS-3828
* https://github.com/riscv/zibi/releases/tag/v0.1.0

topperc commented on llvm/lib/Target/RISCV/RISCVInstrInfoZibimm.td:
```diff
@@ -0,0 +1,48 @@
+//===-- RISCVInstrInfoZvabd.td - 'Zvabd' instructions ------*- tablegen -*-===//
```
Incorrect file name

topperc commented on llvm/lib/Target/RISCV/MCTargetDesc/RISCVMCCodeEmitter.cpp:
```diff
@@ -400,6 +404,22 @@ RISCVMCCodeEmitter::getImmOpValueAsr1(const MCInst &MI, unsigned OpNo,
+
+  if (MO.isImm()) {
+    uint64_t Res = MO.getImm();
+    if (Res >= 1 && Res <= 31)
```
You only need to check for the -1 case. For any other value you can just return Res. The creator of the MCInst is required to make sure that the only possible values are -1, 1-31.

topperc commented on llvm/lib/Target/RISCV/MCTargetDesc/RISCVMCCodeEmitter.cpp:
```diff
@@ -400,6 +404,23 @@ RISCVMCCodeEmitter::getImmOpValueAsr1(const MCInst &MI, unsigned OpNo,
+    return Res;
+  }
+
+  return getImmOpValue(MI, OpNo, Fixups, STI);
```
Will we ever get here? If it's not an immediate we don't have a relocation to encode the cimm correctly. So I this isImm must always be true?

topperc commented on llvm/test/CodeGen/RISCV/zibimm.ll:
```diff
@@ -0,0 +1,299 @@
+; NOTE: Assertions have been autogenerated by utils/update_llc_test_checks.py
+; RUN: llc -mtriple=riscv32 -mattr=+experimental-zibimm -verify-machineinstrs < %s \
+; RUN:     | FileCheck -check-prefix=RV32I-ZIBIMM %s
```
Use a common check prefix for RV32 and RV64 since the code is identical for all cases.

topperc commented:
Please update RISCVUsage.rst and the release notes

BoyaoWang430 commented:
> > This adds the support of Zibi v0.1 experimental extension.
> > References:
> > 
> > * https://lf-riscv.atlassian.net/wiki/spaces/USXX/pages/599261201/Branch+with+Immediate+Zibi+Ratification+Plan
> 
> I cannot for the life of me find an actual specification linked from that page, nor from searching. Where is the GitHub repo in the riscv organisation with it?

There’s no public GitHub repo for Zibi yet. It’s currently going through the ratification process, the repo hasn’t been created at this stage. Please refer to the attachment in [RVS-3828](https://lf-riscv.atlassian.net/browse/RVS-3828) for more details.

topperc commented:
> > > This adds the support of Zibi v0.1 experimental extension.
> > > References:
> > > 
> > > * https://lf-riscv.atlassian.net/wiki/spaces/USXX/pages/599261201/Branch+with+Immediate+Zibi+Ratification+Plan
> > 
> > 
> > I cannot for the life of me find an actual specification linked from that page, nor from searching. Where is the GitHub repo in the riscv organisation with it?
> 
> Sorry, we are still working on the repo. This extension has been approved as fast-track, but we are not familiar with the process (and the process has some changes recently). You may refer to the mailing list (https://lists.riscv.org/g/sig-scalar-efficiency/topic/111893156). We will make all the resouces publicly accessible ASAP, sorry for the inconvenience.

Note the pdf linked there seems to be missing the image that contains the immediate format.

lenary commented on llvm/docs/RISCVUsage.rst:
```diff
@@ -326,6 +326,9 @@ The primary goal of experimental support is to assist in the process of ratifica
 ``experimental-zalasr``
   LLVM implements the `0.0.5 draft specification <https://github.com/mehnadnerd/riscv-zalasr>`__.
 
+``experimental-zibi``, ``experimental-zibi``
```
```suggestion
``experimental-zibi``
```

lenary reviewed (approved):
LGTM with one tiny nit.

wangpc-pp commented on llvm/lib/Target/RISCV/RISCVInstrInfoZibi.td:
```diff
@@ -0,0 +1,51 @@
+
+class Branch_imm<bits<3> funct3, string opcodestr>
+    : RVInstBIMM<funct3, OPC_BRANCH, (outs),
+              (ins GPR:$rs1, uimm5_zibi:$cimm, bare_simm13_lsb0:$imm12),
```
Indent here.

wangpc-pp reviewed (commented):
Please separate this PR into two commits: one for MC and another for CodeGen.

BoyaoWang430 committed:
[RISCV][MC] Add MC support of Zibi experimental extension

This adds the MC support of Zibi v0.1 experimental extension.

References:
https://lf-riscv.atlassian.net/wiki/spaces/USXX/pages/599261201/Branch+with+Immediate+Zibi+Ratification+Plan
https://lf-riscv.atlassian.net/browse/RVS-3828

topperc commented on llvm/lib/Target/RISCV/RISCVInstrInfoZibi.td:
```diff
@@ -0,0 +1,42 @@
+/// This file describes the RISC-V instructions for 'Zibi' (branch with imm).
+///
+//===----------------------------------------------------------------------===//
+// A 5-bit unsigned immediate representing 1-31 and -1. 00000 represents -1.
```
Add a blank line here

topperc commented on llvm/lib/Target/RISCV/RISCVInstrInfoZibi.td:
```diff
@@ -0,0 +1,42 @@
+  let mayLoad = 0;
+  let mayStore = 0;
+}
+let Predicates = [HasStdExtZibi] in {
```
Add a blank line here

jrtc27 commented:
> > > This adds the support of Zibi v0.1 experimental extension.
> > > References:
> > > 
> > > * https://lf-riscv.atlassian.net/wiki/spaces/USXX/pages/599261201/Branch+with+Immediate+Zibi+Ratification+Plan
> > 
> > 
> > I cannot for the life of me find an actual specification linked from that page, nor from searching. Where is the GitHub repo in the riscv organisation with it?
> 
> There’s no public GitHub repo for Zibi yet. It’s currently going through the ratification process, the repo hasn’t been created at this stage. Please refer to the attachment in [RVS-3828](https://lf-riscv.atlassian.net/browse/RVS-3828) for more details.

I’m not sure we should be merging support for extensions that are so early in the development process that they don’t even have a GitHub repository in the riscv organisation with a specification.

BoyaoWang430 committed:
Format adjustment and some bugfix

wangpc-pp commented on llvm/lib/Target/RISCV/RISCVInstrInfoZibi.td:
```diff
@@ -0,0 +1,44 @@
+let Predicates = [HasStdExtZibi] in {
+  def BEQI : Branch_imm<0b010, "beqi">;
+  def BNEI : Branch_imm<0b011, "bnei">;
+} // Predicates = [HasStdExtZibi]
```
Add an blank line at the end of file.

wangpc-pp reviewed (approved):
LGTM. We will hold this PR for one more day to see if there are more comments.

arichardson reviewed (commented):
It would be nice to see an actual encodings spec PDF to be able to verify that the opcode encoding in this PR is actually correct.

The PDF I downloaded from JIRA did not seem to have any encodings.

Two minor comments inline.

wangpc-pp commented:
> It would be nice to see an actual encodings spec PDF to be able to verify that the opcode encoding in this PR is actually correct.

Thanks @arichardson ! We will release an v0.1 PDF in https://github.com/riscv/zibi this week and then update this PR. :-)

topperc commented on llvm/lib/Target/RISCV/RISCVFeatures.td:
```diff
@@ -78,6 +78,12 @@ def FeatureStdExtE
+def FeatureStdExtZibi
+    : RISCVExperimentalExtension<0, 1, "Branch with Immediate">;
+def HasStdExtZibi : Predicate<"Subtarget->hasStdExtZibi()">,
+                      AssemblerPredicate<(all_of FeatureStdExtZibi),
```
indented 2 spaces too far

BoyaoWang430 committed:
Some format adjustment

BoyaoWang430 committed:
Update zibi release specification

BoyaoWang430 commented:
Ping. v0.1 PDF has been released in https://github.com/riscv/zibi/releases/tag/v0.1.0

lenary reviewed (approved):
LGTM. Thanks for publishing a spec.

BoyaoWang430 committed:
change uimm5_zibi to imm5_zibi

topperc reviewed (approved):
LGTM

arichardson reviewed (approved):
Thanks for addressing my comment. The LLVM change looks good to me.

Not related to LLVM but Zibi in general: Since this is the last available branch encoding, I am slightly concerned that we are basing the immediate allocation of this on static instruction count in SPEC2017. Could we gain more code density by having a slightly more complicated immediate? The PDF details the number of bits needed to convert a given percentage of branches, but I imagine certain values (e.g. multiples of 16?) are more common.
I wonder if something slightly more complicated could match more immediates?

For example in the CHERI spec for [YBNDSWI we use one bit to indicate whether to shift the immediate by 4](https://riscv.github.io/riscv-cheri/#SCBNDSI).
Would something like the current format work but with the immediate reduced to 4 bits and MSB being a "shift by 8/16/32"? flag? So you could represent `-1`, `-8`, `1-16`, `(1-16) << 8` instead of `-1` and `1-32`? I can't imagine numbers like 17 being super common. Anyway I will take this comment to the relevant RISC-V mailing list.
